### PR TITLE
Implement sphinx key from v2 config

### PR DESCRIPTION
**Pull request by stsewd**
This PR on top of #4456. It should be merged after.

New commits are from  6756b7f

Ref https://github.com/rtfd/readthedocs.org/issues/4219

### CODE REVIEW

**Comment by humitos** [comment]
I will wait until the base PR gets merged to review this.

**Comment by stsewd** [comment]
Rebased and updated! I'll try to manually test this by morning monday

**Comment by stsewd** [comment]
Ok, I need to figure out how to keep the previous behavior that creates a conf.py file if no one is found and use that.

**Comment by stsewd** [comment]
I'll like to merge this first https://github.com/rtfd/readthedocs.org/pull/4566, as I'm touching some related code and I want to make sure I'm not breaking that code.

File: readthedocs/rtd_tests/tests/test_doc_builder.py [hunk 1]
```diff
@@ -48,7 +48,8 @@ def setUp(self):
     @patch('readthedocs.doc_builder.backends.sphinx.BaseSphinx.run')
     @patch('readthedocs.builds.models.Version.get_conf_py_path')
     @patch('readthedocs.builds.models.Project.conf_file')
-    def test_create_conf_py(self, conf_file, get_conf_py_path, _, get_config_params, create_index, docs_dir):
+    @patch('readthedocs.projects.models.Project.checkout_path')
+    def test_create_conf_py(self, checkout_path, conf_file, get_conf_py_path, _, get_config_params, create_index, docs_dir):
```
Comment (stsewd) [inline readthedocs/rtd_tests/tests/test_doc_builder.py]: Silly me, left this test with a duplicated name...

File: readthedocs/rtd_tests/tests/test_doc_builder.py [hunk 2]
```diff
@@ -86,8 +102,9 @@ def test_create_conf_py(self, conf_file, get_conf_py_path, _, get_config_params,
     @patch('readthedocs.doc_builder.backends.sphinx.BaseSphinx.get_config_params')
     @patch('readthedocs.doc_builder.backends.sphinx.BaseSphinx.run')
     @patch('readthedocs.builds.models.Version.get_conf_py_path')
-    def test_create_conf_py(
-            self, get_conf_py_path, _, get_config_params,
+    @patch('readthedocs.projects.models.Project.checkout_path')
+    def test_multiple_conf_py(
```
Comment (stsewd) [inline readthedocs/rtd_tests/tests/test_doc_builder.py]: Silly me, left this test with a duplicated name...

File: readthedocs/rtd_tests/tests/test_doc_builder.py [hunk 3]
```diff
@@ -97,14 +114,24 @@ def test_create_conf_py(
         """
 
         tmp_docs_dir = py.path.local(tempfile.mkdtemp())
-        tmp_docs_dir.join('conf.py').new()
-        tmp_docs_dir.join('test').mkdir().join('conf.py').new()
+        tmp_docs_dir.join('conf.py').write('')
```
Comment (stsewd) [inline readthedocs/rtd_tests/tests/test_doc_builder.py]: Looks like new() doesn't create the files on disk

**Comment by agjohnson** [comment]
#4456 is merged, this needs a rebase.

File: readthedocs/doc_builder/backends/sphinx.py [hunk 4]
```diff
@@ -151,14 +155,15 @@ def get_config_params(self):
...
+        if self.config.sphinx.configuration is None:
             master_doc = self.create_index(extension='rst')
             self._write_config(master_doc=master_doc)
 
         try:
-            outfile_path = self.project.conf_file(self.version.slug)
+            outfile_path = (
+                self.config.sphinx.configuration or
+                self.project.conf_file(self.version.slug)
+            )
```
Comment (agjohnson) [inline readthedocs/doc_builder/backends/sphinx.py]: It seems here we're implying that if you set a sphinx configuration file, but it doesn't exist, we just create it? I feel like we should error out if you set up a config file and it doesn't actually exist, it will be confusing to the user for the build to pass in this case.

File: readthedocs/doc_builder/backends/sphinx.py [hunk 5]
```diff
@@ -209,9 +216,15 @@ def build(self):
             '.',
             self.sphinx_build_dir,
         ])
+        config_file = (
+            self.config.sphinx.configuration or
+            self.project.conf_file(self.version.slug)
+        )
```
Comment (agjohnson) [inline readthedocs/doc_builder/backends/sphinx.py]: Same here. And I feel like we shouldn't be guessing here. If the config file was specified and it exists, that's when we should set the config file path. Likewise, if we're generating a config, we set a config file path when we find out the user doesn't have a config and didn't specify a config file path.

This will cut down on magical decisions that tend to hide errors from us.

File: readthedocs/doc_builder/backends/sphinx.py [hunk 6]
```diff
@@ -347,7 +360,9 @@ class PdfBuilder(BaseSphinx):
 
     def build(self):
         self.clean()
-        cwd = self.project.conf_dir(self.version.slug)
+        config_file = self.config.sphinx.configuration
+        config_dir = os.path.dirname(config_file) if config_file else None
+        cwd = config_dir or self.project.conf_dir(self.version.slug)
```
Comment (agjohnson) [inline readthedocs/doc_builder/backends/sphinx.py]: This logic looks duplicated, is there a way to reduce this duplication if so?

File: readthedocs/doc_builder/config.py [hunk 7]
```diff
@@ -31,7 +31,10 @@ def load_yaml_config(version):
     python_version = 3 if project.python_interpreter == 'python3' else 2
     allow_v2 = project.has_feature(Feature.ALLOW_V2_CONFIG_FILE)
     try:
-        sphinx_configuration = version.get_conf_py_path()
+        sphinx_configuration = path.join(
+            version.get_conf_py_path(),
+            'conf.py'
+        )
```
Comment (agjohnson) [inline readthedocs/doc_builder/config.py]: It's not clear to me why this change is needed. Shouldn't the full path have already been returned by `version.get_conf_py_path()`?

File: readthedocs/doc_builder/backends/sphinx.py (hunk 4 above)
Comment (stsewd) [reply readthedocs/doc_builder/backends/sphinx.py]: This is to keep compatibility with the behavior when the user doesn't set a config file and rtd tries to find one.

File: readthedocs/doc_builder/backends/sphinx.py (hunk 6 above)
Comment (stsewd) [reply readthedocs/doc_builder/backends/sphinx.py]: I knew you were not going to like it p: haha, so all this is to make sure that we are keeping the previous behavior. If a config file doesn't exist, we create one, and also we try to find one in

File: readthedocs/doc_builder/config.py (hunk 7 above)
Comment (stsewd) [reply readthedocs/doc_builder/config.py]: I was bitten by the name here too, this only returns the directory name. There is a test unskipped that covers this case.

File: readthedocs/doc_builder/backends/sphinx.py (hunk 4 above)
Comment (agjohnson) [reply readthedocs/doc_builder/backends/sphinx.py]: That part I understand, but this seemed to be creating a 3rd option, where the config was specified but not found, and therefore the config was generated instead. This might not be an issue, I interpreted `self.config.sphinx.configuration` as defaulting to `None` if the file is specified but doesn't exist, but that doesn't seem to be the logic used looking at it again.

File: readthedocs/doc_builder/config.py (hunk 7 above)
Comment (stsewd) [reply readthedocs/doc_builder/config.py]: Also currently we are reading this from the database, that's why the builds don't fail

**Comment by stsewd** [comment]
V1 is still working, I refactored the code. I'll be testing v2 in the morning.

**Comment by stsewd** [comment]
OK, v2 tested! It works :D

**Comment by agjohnson** [comment]
Changes look good!
